### ISSUE TITLE
Request: add UCI option, "variety"

### ISSUE THREAD

**Issue by 0zymandia2** [lantonov/asmFish#60]
Komodo has offered this since version 10.4.

Kai Lassos did some testing with this option, and it looks like SF might benefit from it [even more than Komodo](http://www.talkchess.com/forum/viewtopic.php?p=709073#709073). It would be helpful to avoid doubles at hyper-bullet TCs.

Unless I missed something, this hasn't been brought up before.
PS: tried it at the official branch with no reply:  https://github.com/official-stockfish/Stockfish/issues/1131

**Comment by 0zymandia2** [comment]
With the help of Michael Byrne, I could modify the [latest dev](http://www26.zippyshare.com/v/L8TjB6Fw/file.html), timestamped 1498079165 . BTW, this is the only patch worth mentioning, that still hasn't made it's way to asmFish. There's been almost no ELO improvement for more than a month now.
Only a few lines of code added to the files search.cpp and ucioption.cpp, in the mentioned dev, to activate the variety feature.

**Comment by hero2017** [comment]
I'd surely be happy to see this feature and set the default to whatever SF is without this option.

________________________________

**Comment by 0zymandia2** [comment]
The default value, to work as SF would without this option, is 0, which is the one currently enabled.

If you do some testing let's us know, I've seen disconnects when not having the syzygy DTZ files around, I won't be able to test properly for several days.

Edit: tested with the latest cutechess-cli binaries, no problems with the whole 6-men syzygy set available. Even at a value of 8, the ELO drop is very small, almost between the margin of error.

**Comment by lantonov** [comment]
Sorry I was late in answering. It's because I was waiting to have the opinion of SF maintainers. Their reason for refusal (disregarding feature requests) does not pertain completely for asmFish because we do mind some feature requests. 
Nevertheless, I am a bit reluctant to introduce some form of "variety" to asmFish even as an option because it means to deliberately increase the indeterminacy (randomness) in evaluation. Randomness is a bad thing to have and the effort is to achieve more deterministic evaluation. In this regard, Stockfish is better than Komodo having a more deterministic evaluation as seen from the thread that you gave in your first post (Kai Laskos). There was even a special patch last November that fixed issue [#859](https://github.com/official-stockfish/Stockfish/issues/859) connected with indeterminacy. 
Even for the casual user, there is no need to deliberately introduce more randomness. There are 2 big sources of randomness in evaluation which guarantee that there will not be repeating games from any given position: (1) when playing on time controls, there are randomly differing time intervals for each move and (2) when playing with more than 1 thread, there is random selection of threads and depths for each variation.

**Comment by 0zymandia2** [comment]
> I am a bit reluctant to introduce some form of "variety" to asmFish even as an option because it means to deliberately increase the indeterminacy (randomness) in evaluation. Randomness is a bad thing to have and the effort is to achieve more deterministic evaluation.
There are 2 big sources of randomness in evaluation which guarantee that there will not be repeating games from any given position: (1) when playing on time controls, there are randomly differing time intervals for each move and (2) when playing with more than 1 thread, there is random selection of threads and depths for each variation.

Could those among us (I'm guessing I'm not alone) who test at fixed depth & 1 thread, have the option to assemble asmFish with this feature? It could be left out, for the standard version (0) and only available for those who wish to toy with it (1). The same way we had pedantic and "regular" before, we could have official (deterministic) and random (with the UCI option, "variety" coded).

**Comment by Zerbinati** [comment]
It would be possible to convert the c ++ code into assembly for us
And provide us with an unofficial version?
There are only a few lines of code and we will be very grateful to you.
search.cpp
//Add line 145
int variety;
// Add line 248
variety			= Options["Variety"];
//Add a little variety to play (Add line 1337)
		if (variety && value + (variety * 5 * PawnValueEg / 100) >= 0 )
			value += rand() % (variety * 5);
ucioption.cpp
// Add line 65
o["Variety"]               << Option(0, 0, 8);

**Comment by tthsqe12** [comment]
This kind of request cannot be added to the default version of asmfish because the default value variety=0 is going to be a slow down. Also, it would not be nice to ask Lantanov to make executables for a special version (and every other version that gets requested). Therefore, the best solution is to add an assemble option like the current USE_BOOK, USE_WEAKNESS, ..., have you manually change "USE_VARIETY equ 0" to "USE_VARIETY equ 1" in the file you want, and have you assemble it for yourself.


This commit should do what you want
https://github.com/tthsqe12/asmFish/commit/65979b9c49e281e3049316ab9c970ced7ffe6b54
If lantanov like this, than it will get moved from my fork.

Note, I do not know why you like the variety in multiples of five, so I have changed you code slightly to
```
o["Variety"] << Option(0, 0, 40);

.
.
.

       }
    }

  if (bestValue + (variety * PawnValueEg / 100) >= 0 )
    bestValue += rand() % (variety + 1); 

    // All legal moves have been searched. A special case: If we're in check
    // and no legal moves were found, it is checkmate.
    if (InCheck && bestValue == -VALUE_INFINITE)
        return mated_in(ss->ply); // Plies to mate from the root

    tte->save(posKey, value_to_tt(bestValue, ss->ply),
              PvNode && bestValue > oldAlpha ? BOUND_EXACT : BOUND_UPPER,
              ttDepth, bestMove, ss->staticEval, TT.generation());

    assert(bestValue > -VALUE_INFINITE && bestValue < VALUE_INFINITE);

    return bestValue;
}

**Comment by Zerbinati** [comment]
> Very kind
Thank you very much!

**Comment by Ipmanchess** [comment]
Great idea.. request adding by selection 0 or 1 and assemble yourself!!

**Comment by 0zymandia2** [comment]
> This kind of request cannot be added to the default version of asmfish because the default value variety=0 is going to be a slow down. Also, it would not be nice to ask Lantanov to make executables for a special version (and every other version that gets requested). Therefore, the best solution is to add an assemble option like the current USE_BOOK, USE_WEAKNESS, ..., have you manually change "USE_VARIETY emu 0" to "USE_VARIETY emu 1" in the file you want, and have you assemble it for yourself.

Yes, that's exactly what I proposed in my [previous comment](https://github.com/lantonov/asmFish/issues/60#issuecomment-311951609), maybe I didn't make myself clear.

**Comment by lantonov** [comment]
Thanks, @tthsqe12 !! You are always so kind although it takes a portion of your valuable time. Gladly adding this option.

**Comment by hero2017** [comment]
First, I wanted to thank you tthsque12 for adding this wonderful feature. Your time and efforts put into the best chess engine IMHO is greatly appreciated!

However, on my dual E5-2696v3 machine, at least with HT enabled, I get about a 50% loss in speed. That's huge and totally unexpected. I don't know why this would cause this and since nobody else mentioned I assume it's just my machine. Any idea why there would be such a huge loss in speed simply by enabling this?

I haven't had time to disable HT yet to see if it makes any difference but when I do I will add my findings here.

Using Win'2016 Server.

**Comment by lantonov** [comment]
Here are the results from BuildTester
```
Build Tester: 1.4.6.0
Windows 10 (Version 10.0, Build 0, 64-bit Edition)
Intel(R) Core(TM) i5-4210U CPU @ 1.70GHz
SafeMode: No 
Running In VM: No 
HyperThreading Enabled: No 
CPU Warmup: Yes
Command Line: bench 256 4 18 default depth
Tests per Build: 30
ANOVA: n/a

                Engine# (NPS)                     Speedup     Sp     Conf. 95%    S.S.
1  (4,406,447.8   ) ---> 2  (4,364,741.9   ) --->  0.956%  59,698.7     Yes       No 
```
Non-variety engine (No. 1) has less than 1% more speed than variety engine (No. 2) which is not significant at 95% confidence.

**Comment by hero2017** [comment]
That may be true on that machine, but not on a dual E5-2696v3 (72 cores).

**Comment by lantonov** [comment]
Would you give some speed comparison stats (FishBench, BuildTester) on your 72-core machine? With and without HT

**Comment by hero2017** [comment]
Build Tester: 1.4.7.0
Windows 10 (Version 10.0, Build 0, 64-bit Edition)
Intel(R) Xeon(R) CPU E5-2696 v3 @ 2.30GHz
SafeMode: No 
Running In VM: No 
HyperThreading Enabled: Yes
CPU Warmup: Yes
Command Line: bench 256 4 18 default depth
Tests per Build: 30
ANOVA: n/a

                Engine# (NPS)                     Speedup     Sp     Conf. 95%    S.S.
2  (8,936,891.6   ) ---> 1  (8,466,710.5   ) --->  5.553%  218,613.6    Yes       No 

At first I didn't see much difference in speed using BenchTester, only 5.5% with HT enabled...but then I realized it wasn't using all my cores. I figured the 4 after bench 256 means cores so I changed it to 71, leaving 1 core for the bench (what does the 18 mean? I figure the 256 means ram).

So I left HT enabled and everything else the same but this time I changed the 4 to 71 (cores) and you can see the big difference:

```
Build Tester: 1.4.7.0
Windows 10 (Version 10.0, Build 0, 64-bit Edition)
Intel(R) Xeon(R) CPU E5-2696 v3 @ 2.30GHz
SafeMode: No 
Running In VM: No 
HyperThreading Enabled: Yes
CPU Warmup: Yes
Command Line: bench 256 71 18 default depth
Tests per Build: 30
ANOVA: n/a

                Engine# (NPS)                     Speedup     Sp     Conf. 95%    S.S.
2  (88,843,801.1  ) ---> 1  (47,821,728.5  ) ---> 85.781%  1,728,193.6    Yes       No 
```

I was guessing 50% slower but I just did that calc quickly in my head where the speed was just being reported by the gui (Aquarium) and obviously I miscalculated. The difference is actually 85% slower with variety set to 1 before compiling (I just use make.bat after).

**Comment by hero2017** [comment]
Now with HT disabled and 35 cores:

```
Build Tester: 1.4.7.0
Windows 10 (Version 10.0, Build 0, 64-bit Edition)
Intel(R) Xeon(R) CPU E5-2696 v3 @ 2.30GHz
SafeMode: No 
Running In VM: No 
HyperThreading Enabled: Yes
CPU Warmup: Yes
Command Line: bench 256 35 18 default depth
Tests per Build: 30
ANOVA: n/a

                Engine# (NPS)                     Speedup     Sp     Conf. 95%    S.S.
2  (70,383,571.6  ) ---> 1  (29,403,613.3  ) ---> 139.370%  1,108,116.8    Yes       No 
```

Btw, I did enable warmup cpu for each test.

Well, this time the results are even worse almost 140% ???!! Wow!

**Comment by hero2017** [comment]
One more thing which perhaps will help you hunt down the reason. With Variety enabled on another machine (dual X5670) and HT off I didn't see a significant loss in speed. The main difference between these two machines is that the X5670 I used the popcnt version whereas for the e5-2696v3 I used bmi2. However, that's just what I observed in Aquarium and I didn't run any bench tests on it. No time right now.

**Comment by lantonov** [comment]
Wow, that's huge! Seems like HT is the culprit

**Comment by hero2017** [comment]
Why HT? With HT disabled it's even worse (140% slower compared to 85% with HT enabled).

**Comment by lantonov** [comment]
The above output shows that both 140% and 85% are with HT enabled
`HyperThreading Enabled: Yes`

**Comment by hero2017** [comment]
I didn't notice that. In that case it's reporting it incorrectly (not detecting HT off). I definitely disabled HT for that test and it was only 35 cores.

**Comment by hero2017** [comment]
I had a few mins so I tried it again:

```
Windows PowerShell
Copyright (C) 2016 Microsoft Corporation. All rights reserved.

PS C:\Users\hero2017.mydomain> Get-WmiObject -class Win32_Processor | ft Name,DeviceID,NumberOfCores,NumberOfLogicalProcessors

Name                                      DeviceID NumberOfCores NumberOfLogicalProcessors
----                                      -------- ------------- -------------------------
Intel(R) Xeon(R) CPU E5-2696 v3 @ 2.30GHz CPU0                18                        18
Intel(R) Xeon(R) CPU E5-2696 v3 @ 2.30GHz CPU1                18                        18

Powershell reports it correctly, 36 cores so HT is off.

I ran BenchTester again and set it to 35 cores and I got the same result:

Build Tester: 1.4.7.0
Windows 10 (Version 10.0, Build 0, 64-bit Edition)
Intel(R) Xeon(R) CPU E5-2696 v3 @ 2.30GHz
SafeMode: No 
Running In VM: No 
HyperThreading Enabled: Yes
CPU Warmup: Yes
Command Line: bench 256 35 18 default depth
Tests per Build: 30
ANOVA: n/a

                Engine# (NPS)                     Speedup     Sp     Conf. 95%    S.S.
2  (70,427,709.3  ) ---> 1  (29,546,417.6  ) ---> 138.363%  1,079,138.9    Yes       No 

I can confirm BenchTester is definitely reporting HT on when it's really off.
```

**Comment by hero2017** [comment]
Windows PowerShell
Copyright (C) 2016 Microsoft Corporation. All rights reserved.

PS C:\Users\hero2017.mydomain> Get-WmiObject -class Win32_Processor | ft Name,DeviceID,NumberOfCores,NumberOfLogicalProcessors

Name                                      DeviceID NumberOfCores NumberOfLogicalProcessors
----                                      -------- ------------- -------------------------
Intel(R) Xeon(R) CPU E5-2696 v3 @ 2.30GHz CPU0                18                        36
Intel(R) Xeon(R) CPU E5-2696 v3 @ 2.30GHz CPU1                18                        36

Now you can see HT is on, 72 cores...and I ran BenchTester again this time with 72 cores and the result is similar to before:

Build Tester: 1.4.7.0
Windows 10 (Version 10.0, Build 0, 64-bit Edition)
Intel(R) Xeon(R) CPU E5-2696 v3 @ 2.30GHz
SafeMode: No 
Running In VM: No 
HyperThreading Enabled: Yes
CPU Warmup: Yes
Command Line: bench 256 72 18 default depth
Tests per Build: 30
ANOVA: n/a

                Engine# (NPS)                     Speedup     Sp     Conf. 95%    S.S.
1  (88,976,364.9  ) ---> 2  (48,978,115.4  ) ---> 81.666%  1,847,682.3    Yes       No 

![image](https://user-images.githubusercontent.com/20324387/27997637-e50d26a6-64ca-11e7-85a7-1f6fe612ffce.png)

**Comment by lantonov** [comment]
Ok, but if not HT then what? Looking at the variety patch, I see two pieces of code that might cause this: one in QSearchMacros.asm (more probable) and the other in Uci.asm. But I am not the specialist, @tthsqe12 knows better.

**Comment by hero2017** [comment]
I'm going to keep using asmF without Variety...maybe when @tthsqe12 has some spare time he can look into it.

**Comment by lantonov** [comment]
If you want to have a strong program (many cores, high speed, etc.) you should avoid randomness ("variety"). Randomness and precision in evaluation are incompatible.

**Comment by hero2017** [comment]
Yes I think so too...but I wanted to experiment with it to see if it plays any different or whether it could be useful in some positions.  I'm trying to avoid those dreadful draws as much as possible in correspondence games, especially against weaker opponents...and asmF with contempt=10 isn't helping. Maybe I should try 40 but I'm afraid to do it for real games as I have little computing time for running engine vs engine games.

**Comment by tthsqe12** [comment]
Oh yes! I just realized this is going to be terrible if you have a ton of cores. The problem is that you have a ton of cores competing for the same memory location "options.varietySeed":
https://github.com/lantonov/asmFish/blob/master/asmFish/guts/QSearchMacros.asm#L491#L502

Instead of this, each thread should get its own copy.

**Comment by tthsqe12** [comment]
The commit "use per thread random seeds" in my fork should fix this. If it is still not fast enough, I could consider replacing the idiv's. You can see variety 0 is same as stockfish, then it deviates with 20.
```
$ ./pfish setoption name variety value 0: bench
pedantFishL_2017-07-10_popcnt
*** bench hash 16 threads 1 depth 13 realtime 0 ***
1:      nodes:  206230          1494 knps
2:      nodes:  308133          1721 knps
3:      nodes:  85572           2516 knps
4:      nodes:  179381          1793 knps
5:      nodes:  176351          1617 knps
6:      nodes:  306536          1525 knps
7:      nodes:  264408          1537 knps
8:      nodes:  509077          1749 knps
9:      nodes:  218549          1539 knps
10:     nodes:  161686          1720 knps
11:     nodes:  120033          1600 knps
12:     nodes:  390839          1582 knps
13:     nodes:  234863          1689 knps
14:     nodes:  129305          1576 knps
15:     nodes:  100820          1867 knps
16:     nodes:  87686           1789 knps
17:     nodes:  92085           2192 knps
18:     nodes:  9376            3125 knps
19:     nodes:  122554          2785 knps
20:     nodes:  475256          2475 knps
21:     nodes:  60221           2867 knps
22:     nodes:  42995           3307 knps
23:     nodes:  139947          3588 knps
24:     nodes:  65008           2407 knps
25:     nodes:  7408            1852 knps
26:     nodes:  47027           2351 knps
27:     nodes:  383245          2189 knps
28:     nodes:  262909          1905 knps
29:     nodes:  55479           1913 knps
30:     nodes:  36953           2463 knps
31:     nodes:  61505           3237 knps
32:     nodes:  77538           2982 knps
33:     nodes:  57610           3388 knps
34:     nodes:  111239          2648 knps
35:     nodes:  165345          2624 knps
36:     nodes:  90588           3623 knps
37:     nodes:  415173          2319 knps
===========================
Total time (ms) : 3239
Nodes searched  : 6258930
Nodes/second    : 1932364
$ ./pfish setoption name variety value 20: bench
pedantFishL_2017-07-10_popcnt
*** bench hash 16 threads 1 depth 13 realtime 0 ***
1:      nodes:  291581          1602 knps
2:      nodes:  263578          1780 knps
3:      nodes:  120426          2617 knps
4:      nodes:  142920          1742 knps
5:      nodes:  63463           1762 knps
6:      nodes:  223684          1521 knps
7:      nodes:  217332          1574 knps
8:      nodes:  409125          1771 knps
9:      nodes:  315573          1577 knps
10:     nodes:  214180          1784 knps
11:     nodes:  193864          1576 knps
12:     nodes:  205863          1571 knps
13:     nodes:  258032          1664 knps
14:     nodes:  157907          1595 knps
15:     nodes:  114916          1883 knps
16:     nodes:  75104           1976 knps
17:     nodes:  112759          2301 knps
18:     nodes:  15012           3002 knps
19:     nodes:  53803           2831 knps
20:     nodes:  334181          2531 knps
21:     nodes:  38028           2716 knps
22:     nodes:  34697           3855 knps
23:     nodes:  113169          3143 knps
24:     nodes:  75551           2518 knps
25:     nodes:  14544           2077 knps
26:     nodes:  57865           2411 knps
27:     nodes:  549747          2090 knps
28:     nodes:  171033          1965 knps
29:     nodes:  55190           2044 knps
30:     nodes:  18229           2604 knps
31:     nodes:  67086           3049 knps
32:     nodes:  77734           2989 knps
33:     nodes:  84315           3242 knps
34:     nodes:  129435          2876 knps
35:     nodes:  155751          2639 knps
36:     nodes:  125402          3582 knps
37:     nodes:  585026          2427 knps
===========================
Total time (ms) : 3100
Nodes searched  : 6136105
Nodes/second    : 1979388

**Comment by lantonov** [comment]
Committed "use per thread random seeds" to master in main repo

**Comment by 0zymandia2** [comment]
The whole point of "variety", was to avoid deterministic moves in 1 thread, fixed depth tests. That it **does**, and even at the maximum value of 40, the ELO loss is negligible. Speed is unencumbered, both with the popcnt as well as the bmi2 assembles. HT enabled.

**Comment by lantonov** [comment]
Avoiding deterministic moves in 1 thread, fixed depth tests is easy. The hard problem is to maintain determinism.  I am not sure that Elo loss is negligible. With the latest patch in connection with another issue (#66) determinism in fixed depth was lost, and there was an obvious loss of Elo. With the "Variety" patch, we saw that speed is encumbered to various degrees. Hopefully, though, @tthsqe12 fixed the speed loss problem.

**Comment by hero2017** [comment]
Build Tester: 1.4.7.0
Windows 10 (Version 10.0, Build 0, 64-bit Edition)
Intel(R) Xeon(R) CPU E5-2696 v3 @ 2.30GHz
SafeMode: No 
Running In VM: No 
HyperThreading Enabled: Yes
CPU Warmup: Yes
Command Line: bench 256 72 18 default depth
Tests per Build: 30
ANOVA: n/a

                Engine# (NPS)                     Speedup     Sp     Conf. 95%    S.S.
1  (89,436,580.9  ) ---> 2  (88,957,010.0  ) --->  0.539%  1,306,987.1    No        No 

That fixed it! Thanks to both lantonov & @tthsqe12 for your efforts in maintaining the fastest engine alive.

**Comment by 0zymandia2** [comment]
I don't measure speed by looking at kn/s. I test at fixed depth, so all I do is watch how much CPU time the exe takes, in the task manager. I get the same times there, that's all that matters to me.
With the variety option, I get an engine that plays different moves, and is almost at the same strength.

**Comment by hero2017** [comment]
I don't like this method of measuring speed because this varies. When I run infinite analysis for say 5 minutes it may reach a depth of say 40. Reloading the engine and analyzing the same position up to depth 40 can be 6 mins or 4 minutes. It varies, probably due to some randomness in search. And with Variety on I presume that can vary even more now.

**Comment by 0zymandia2** [comment]
As it turns out, speed is now a moot point, after several hundred thousand games, the original patch caused problems in cutechess-cli. I've tried the more recent one (Jul 10) and it's much worse, after just 50 thousand games, it completely stalled. All the processes were at 0% CPU and after terminating them, cutechess-cli exited without saving the games played to the PGN.

Totally unusable, back to the original (Jul 1), at least that one saved the games upon killing the stalled process (just the one) and resumed work without further incidence.

**Comment by lantonov** [comment]
Do you use large pages?

**Comment by 0zymandia2** [comment]
No.

**Comment by lantonov** [comment]
I have never had problems with AF crashing but I haven't run hundred thousand games either

**Comment by 0zymandia2** [comment]
It never crashed on me before either, at least not that I noticed. I use the -recover switch in cutechess, so I guess most crashes go unnoticed.
Gull used to crash, before basil00's version came around. H5 also crashes (H4 didn't), but only from time to time (once or twice every million games).

**Comment by tthsqe12** [comment]
Are you saying that af becomes unresponsive on your machine? Does this happen without the variety option? Have you noticed this before? The patch that fixed the root move ordering does touch the multithreading code, but I doubt that it could have introduced any bugs.

**Comment by 0zymandia2** [comment]
With cutechess, and using the -recover option, this is what happens with the variety option enabled and at a value of 40. TC fixed depth 11, 1 thread per engine, ponder off hash 16, 6 cores 12 threads, concurrency 12, 6-men syzygy probe depth 1:
After many thousands games, one of the processes stops working but it's not terminated by cutechess, CPU usage gets reduced but it keeps working. Now, if I'm not monitoring the task manager, and I don't kill that process, more processes keep stopping, to the point were all of them are at 0% but then cutechess will be at the maximum CPU usage it can get (8-9% or 1/12 threads). From that point it's troublesome to get it working again, I have to kill the AF processes one by one, waiting for disk activity of the HDD were the OS and the DTZ files are located (not the engines, WDL or cutechess files) to drop from 100%. Every time you kill a process, disk activity will rise to 100% again, for several minutes, until it drops again, that's when cutechess acknowledges the engine failure as a disconnect and adjudicates the game, starting a new process. It's not until you've done this for all 12 engines, that games are finally saved to the PGN and work properly resumed.
This happens bot with the original variety patch from July 1st, as well as the one from July 10th. I'm guessing I killed the processes on that run too quickly and that's why cutechess exited abruptly. Also, the failure rate seems to be on par, after more games were played, so the original one was a lucky run. Both patches are very hard to use in the long run.

**Comment by 0zymandia2** [comment]
This is what the author of the original code said as to why he deleted it from later releases of his SF fork:
http://www.talkchess.com/forum/viewtopic.php?p=720458#720458

Also, check my reply, it looks like the problem might be with the syzygy probing?

**Comment by lantonov** [comment]
Could it be related to CPU's defenses against overheating ?

**Comment by 0zymandia2** [comment]
I use liquid cooling, it never reaches 60ºC

Just checked, room temp 28.5, CPU 54.2, hottest part of the MOBO 57ºC.

**Comment by tthsqe12** [comment]
0zymandia2, I would like to know if you experience the same unresponsiveness from AF without the variety option.
This patch

//Add a little variety to play (Add line 1337)
if (variety && value + (variety * 5 * PawnValueEg / 100) >= 0 )
value += rand() % (variety * 5);

does look a little dodgy since it is possibly returning a score >= VALUE_INFINITE.

This gives undefined behaviour in the rest of the code, and in your case, this undefined behaviour is undesirable.

How about moving the value += rand to the end of evaluate? Due to lazy evaluation and other factors, this is almost certainly going to never get into the mate range.

**Comment by 0zymandia2** [comment]
When you say without the variety option, do you mean setting it to zero or assembling with USE_VARIETY      emu 0?
About moving the value += rand to the end of evaluate, beats me ;-)

**Comment by tthsqe12** [comment]
without variety option === USE_VARIETY equ 0

**Comment by 0zymandia2** [comment]
No, I haven't tried that. Once the current tour is over, I will in a few days.

**Comment by hero2017** [comment]
I never had a problem with CuteChess, but I normally only run a few hundred games. I tried the latest asmF release and CuteChess did not crash, and neither did asmF.

**Comment by 0zymandia2** [comment]
I took advantage of the newly released cutechess-cli 1.0, to perform several tests with the latest asmFish (July 10):

USE_VARIETY equ 0-> no problems
USE_VARIETY equ 1-> no problems leaving the option at default values
USE_VARIETY equ 1-> no problems with the option at the maximum value of 40 (SyzygyProbeLimit 5)
USE_VARIETY equ 1-> problems with the option at the maximum value of 40 (SyzygyProbeLimit 6)

Problems arose quite quickly (around 30K games), syzygy probe depth was lowered to 1 in all cases. TC=fixed depth 11, 1 thread per engine, ponder off, hash 16, 6 cores & 12 threads, concurrency 12, huge bin book for the "variety" asmFish, no book for the other asmFish.

When using variety, speed takes a 15% hit, but strength rises by 13 ELO points. All this at fixed depth 11,

**Comment by tthsqe12** [comment]
this issues is not closed because there are problems in 0zymandia2 last post.

**Comment by tthsqe12** [comment]
0zymandia2, in the command
`value += rand() % (variety * 5);`
do you want the output of rand() to be negative (so that value can go down)?
The current implementation has rand() as signed so that the average change to value is 0.

Also, the issues you experienced might be caused by value overflowing VALUE_INFINITE. Maybe better to put an upper bound as
```
if (variety && value <= SOMEUPPERBOUND && value + (variety * 5 * PawnValueEg / 100) >= 0 )
value += rand() % (variety * 5);

**Comment by 0zymandia2** [comment]
You're talking Chinese to me ;-)

I want variety at a value of 40 to work; just implement the changes you think might solve the problem, and I'll test them. The computer I use for that is idle, waiting for the fix.

**Comment by lantonov** [comment]
@0zymandia2, did you test the variety option that you requested?

**Comment by tthsqe12** [comment]
I haven't implemented the changes that should prevent undefined behaviour. So it is still on hold.

**Comment by lantonov** [comment]
Ok

**Comment by 0zymandia2** [comment]
> So it is still on hold.

Actually, I've been using it, despite the instances stopping it's still faster than the alternative (using another engine). This thing is going to rock, once it works correctly.

**Comment by 0zymandia2** [comment]
I see there's been changes to the use of the variety option. Next week I'm supposed to get a Ryzen 7, I'll test it then.

**Comment by lantonov** [comment]
**From Fishcooking:**
_Michel van der Bergh_
Thx! So it is not only leaf nodes that are perturbed but all nodes (right before they make their way into the tt) in a way which seems incompatible with minimax search (the perturbed bestValue appears to be no longer a minimax value). It is not clear what the implications of this are (probably none).

bestValue += rand() % (variety + 1);
This seems dangerous since each time the position is visited the perturbation will have a different value. There is also a bias in the perturbation (again it is difficult to understand the implications).

Concerning this last point I would still prefer my solution

perturbation=(hashCode^RANDOM)/%(variety+1)-variety/2

(assuming variety is even). RANDOM is a random number which is fixed at the beginning of each game.

**Comment by 0zymandia2** [comment]
What modification does "make.bat" need, to get the equivalent of the old "USE_VARIETY equ 1"?

**Comment by MichaelB7** [comment]
Edit fish.asm change “SetDefault 0, USE_VARIETY” to “SetDefault 1, USE_VARIETY”

**Comment by 0zymandia2** [comment]
Thank you, specially for the original code.

Edit: test completed. No irrecoverable crashes after several million games.
